db/db_batch.go

```diff
@@ -9,34 +9,73 @@ package db
 )
 
-// KVStoreBatch is the interface of Batch KVStore.
-// It buffers the Put/Delete operations and persist to DB in a single transaction upon Commit()
-type KVStoreBatch interface {
-	// Lock locks the batch
-	Lock()
-	// Unlock unlocks the batch
-	Unlock()
-	// Put insert or update a record identified by (namespace, key)
-	Put(string, []byte, []byte, string, ...interface{}) error
-	// PutIfNotExists puts a record only if (namespace, key) doesn't exist, otherwise return ErrAlreadyExist
-	PutIfNotExists(string, []byte, []byte, string, ...interface{}) error
-	// Delete deletes a record by (namespace, key)
-	Delete(string, []byte, string, ...interface{}) error
-	// Size returns the size of batch
-	Size() int
-	// Entry returns the entry at the index
-	Entry(int) (*writeInfo, error)
-	// Clear clear batch write queue
-	Clear() error
-	// Commit commit queued write to db
-	Commit() error
-	// KVStore returns the underlying KVStore
-	KVStore() KVStore
-}
+type (
+	// KVStoreBatch defines a batch buffer interface that stages Put/Delete entries in sequential order
+	// To use it, first start a new batch
+	// b := NewBatch()
+	// and keep batching Put/Delete operation into it
+	// b.Put(bucket, k, v)
+	// b.PutIfNotExists(bucket, k, v)
+	// b.Delete(bucket, k, v)
+	// once it's done, call KVStore interface's Commit() to persist to underlying DB
+	// KVStore.Commit(b)
+	KVStoreBatch interface {
+		// Lock locks the batch
+		Lock()
+		// Unlock unlocks the batch
+		Unlock()
+		// Put insert or update a record identified by (namespace, key)
+		Put(string, []byte, []byte, string, ...interface{}) error
+		// PutIfNotExists puts a record only if (namespace, key) doesn't exist, otherwise return ErrAlreadyExist
+		PutIfNotExists(string, []byte, []byte, string, ...interface{}) error
+		// Delete deletes a record by (namespace, key)
+		Delete(string, []byte, string, ...interface{}) error
+		// Size returns the size of batch
+		Size() int
+		// Entry returns the entry at the index
+		Entry(int) (*writeInfo, error)
+		// Clear clears entries staged in batch
+		Clear() error
+		// internal clear called by Clear()
+		clear() error
+	}
+
+	// writeInfo is the struct to store Put/Delete operation info
+	writeInfo struct {
+		writeType   int32
+		namespace   string
+		key         []byte
+		value       []byte
+		errorFormat string
+		errorArgs   interface{}
+	}
+
+	// baseKVStoreBatch is the base implementation of KVStoreBatch
+	baseKVStoreBatch struct {
+		mutex      sync.RWMutex
+		writeQueue []writeInfo
+	}
+
+	// CachedBatch derives from Batch interface
+	// A local cache is added to provide fast retrieval of pending Put/Delete entries
+	CachedBatch interface {
+		KVStoreBatch
+		// Get gets a record by (namespace, key)
+		Get(string, []byte) ([]byte, error)
+	}
+
+	// cachedBatch implements the CachedBatch interface
+	cachedBatch struct {
+		baseKVStoreBatch
+		cache map[hash.PKHash][]byte // local cache of batched <k, v> for fast query
```

Comment:
We shouldn't call it PKHash. This is confusing. Why database needs to be aware of public key.

Reply:
can rename it, just use it as 20-byte hash